### ISSUE TITLE
memleak test on some builds

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#10715]
The memleak test, on some builds, seems to fail.  See https://travis-ci.org/openssl/openssl/jobs/628227828 for example.

There's a subthread in PR #10689, starting around https://github.com/openssl/openssl/pull/10689#issuecomment-568515995, that talks about failing memleaktest  But now it seems to not be a problem?

I copied the failed config line:
```
l ./config no-asm enable-asan enable-rc5 enable-md2 no-shared -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -D__NO_STRING_INLINES -Wno-unused-command-line-argument
```

Doing a make and running the test:
```
; make V=1 TESTS=test_memleak test
make depend && make _tests
make[1]: Entering directory '/home/rsalz/openssl/github'
make[1]: Leaving directory '/home/rsalz/openssl/github'
make[1]: Entering directory '/home/rsalz/openssl/github'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \
  OPENSSL_MODULES=`cd .././providers 2>/dev/null && pwd` \
  /usr/bin/perl .././test/run_tests.pl test_memleak )
90-test_memleak.t .. 
1..2
/home/rsalz/openssl/github/util/shlib_wrap.sh /home/rsalz/openssl/github/test/memleaktest => 0
not ok 1 - running leak test

#   Failed test 'running leak test'
#   at ../test/recipes/90-test_memleak.t line 18.
/home/rsalz/openssl/github/util/shlib_wrap.sh /home/rsalz/openssl/github/test/memleaktest freeit => 0
ok 2 - running no leak test
# Looks like you failed 1 test of 2.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/2 subtests 

Test Summary Report
-------------------
90-test_memleak.t (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=2,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.02 cusr  0.03 csys =  0.06 CPU)
Result: FAIL
Makefile:2199: recipe for target '_tests' failed
make[1]: *** [_tests] Error 1
make[1]: Leaving directory '/home/rsalz/openssl/github'
Makefile:2197: recipe for target 'tests' failed
```

**Comment by richsalz** [comment]
Running the commands individually works:
```
; ./test/memleaktest ; echo $status        
0
; ./test/memleaktest freeit ; echo $status
0
```

But this is the failing line in 90-test_memleak.t:
```
; cat -n test/recipes/90-test_memleak.t | sed -n 15,20p
    15	    if $^O eq 'darwin';
    16	
    17	plan tests => 2;
    18	ok(!run(test(["memleaktest"])), "running leak test");
    19	ok(run(test(["memleaktest", "freeit"])), "running no leak test");
```

**Comment by richsalz** [comment]
Here is the build command for memleaktest:
```
gcc  -Iinclude -Iapps/include  -pthread -m64 -fsanitize=address  -fno-omit-frame-pointer -g -Wall -O3 -Wno-unused-command-line-argument -DNDEBUG -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -D__NO_STRING_INLINES -MMD -MF test/memleaktest-bin-memleaktest.d.tmp -MT test/memleaktest-bin-memleaktest.o -c -o test/memleaktest-bin-memleaktest.o test/memleaktest.c
touch test/memleaktest-bin-memleaktest.d.tmp
if cmp test/memleaktest-bin-memleaktest.d.tmp test/memleaktest-bin-memleaktest.d > /dev/null 2> /dev/null; then \
	rm -f test/memleaktest-bin-memleaktest.d.tmp; \
else \
	mv test/memleaktest-bin-memleaktest.d.tmp test/memleaktest-bin-memleaktest.d; \
fi
rm -f test/memleaktest
${LDCMD:-gcc} -pthread -m64 -fsanitize=address  -fsanitize=leak -fno-omit-frame-pointer -g -Wall -O3 -Wno-unused-command-line-argument -L.   \
	-o test/memleaktest \
	test/memleaktest-bin-memleaktest.o \
	test/libtestutil.a -lcrypto -ldl -pthread 
; 
```
This seems to be wrong; looking at https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html it seems that ```-fsanitize=leak``` is needed on the link command. When I add that flag by hand (doing 'make -n' into a script, editing it, and then running the script) I still don't get failures using gcc (version 7.4.0 on ubuntu).  If I change the gcc to clang, it *does* work.

So, long story short: perhaps just use clang for leak-checking?  @bernd-edlinger are we using the wrong version of gcc to enable this?

**Comment by bernd-edlinger** [comment]
I am unable to reproduce, for me the test always works, but I use something gcc-10 built locally.
I never use -fsanitize=leak, I think it is automatically enabled.
Which gcc version is this exactly?

**Comment by bernd-edlinger** [comment]
And, BTW, this appears to be failing with clang-7.0.0, right?:

> The memleak test, on some builds, seems to fail. See https://travis-ci.org/openssl/openssl/jobs/628227828 for example.

**Comment by richsalz** [comment]
I can get it to fail on gcc pass on clang with these versions on my machine:
```
; clang --version
clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
; gcc --version
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

;
```

**Comment by richsalz** [comment]
And to get those pass/fail results, I have to use the specific config line mentioned up in the initial comment here.

**Comment by bernd-edlinger** [comment]
Is this a 64bit build or a 32bit?

**Comment by bernd-edlinger** [comment]
BTW: I can make it fail if I do for instance this:
`LSAN_OPTIONS=detect_leaks=0 make test TESTS=test_memleak V=1`
so are LSAN_OPTIONS or ASAN_OPTIONS defined?

**Comment by bernd-edlinger** [comment]
Still what I don't understand, it seems that the build with the problem in the travis is
explicitly using clang.

**Comment by richsalz** [comment]
The test matrix tries to set the env thing, but perhaps it's not working right or we're confused?  I wonder if puttingg
```
$ENV{LSAN_OPTIONS}='detect_leaks=1;'
```
explicitly in the recipes/90-test_memleak.t would help?

**Comment by bernd-edlinger** [comment]
I am still confused.  Can you give me some more background.
We have a travis build failing 100% in the memleak test, right, it uses clang?
You have been able to reproduce a failure with gcc on your computer?
it does target a 64-bit x86_64 linux?
I am asking, because gcc-7.x does not detect leaks in 32-bit targets, only 64-bit.
The gcc-10 will have leak sanitizer also for 32-bit targets.

**Comment by richsalz** [comment]
I'm confused too.  I'll try to restate things.  First, I agree that the build log that fails is using clang; I did not realize that. This makes things even more confusing.

When I copied the config line from the failing build log, gcc failed the memleak test on my machine.  I did "touch test/memleak.c ; make -n test/memleak >foo" and edited the script to use clang it and worked.

I guess this means that I cannot reproduce the failure except in slightly different circumstances.

**Comment by bernd-edlinger** [comment]
I fact, it can be challenging to debug something that is falling only on travis.
Was the failure rate on the travis build at 100% ?

**Comment by richsalz** [comment]
Perhaps @mattcaswell knows the failure rate; I don't.

**Comment by p-steuer** [comment]
I was not aware of this issue, the failures i see are probably related: https://github.com/openssl/openssl/issues/10870

**Comment by t8m** [comment]
Closing as duplicate of #10870.
